evalcache.plugin.zsh

```diff
@@ -17,18 +17,26 @@ function _evalcache () {
 
   local cacheFile="$ZSH_EVALCACHE_DIR/init-${1##*/}-${cmdHash}.sh"
 
+  local name
+  for i in {1..$#}; do
+    name=${@:$i:1}
+    if [[ $name != *'='* ]]; then
+      break
+    fi
+  done
+
   if [ "$ZSH_EVALCACHE_DISABLE" = "true" ]; then
     eval "$("$@")"
   elif [ -s "$cacheFile" ]; then
     source "$cacheFile"
   else
-    if type "$1" > /dev/null; then
-      (>&2 echo "$1 initialization not cached, caching output of: $*")
+    if type "$name" > /dev/null; then
+      (>&2 echo "$name initialization not cached, caching output of: $*")
       mkdir -p "$ZSH_EVALCACHE_DIR"
-      "$@" > "$cacheFile"
+      eval "$*" > "$cacheFile"
```

Comment:
I don't think `$*` will work here for args with spaces, e.g.:
```zsh
$ echo 1 > a
$ echo 2 > "b c"
$ echo 3 > d

$ function eval_args() { eval "$*" }

$ eval_args cat a "b c" d
1
cat: b: No such file or directory
cat: c: No such file or directory
3

$ eval_args X=1 Y=2 printenv X Y
1
2
```

perhaps something like this approach might work?
```zsh
$ function eval_args () {
  local ARG
  ENV=()
  for ARG in $@; do
    if [[ ${ARG} =~ "^\w+=" ]]; then
      ENV+=(${ARG})
      shift
    else
      break
    fi
  done
  env ${ENV} $@
}

$ eval_args cat a "b c" d
1
2
3

$ eval_args X=1 Y=2 printenv X Y
1
2
```

Reply:
No worries! I have a PR of my own and was just seeing whether it would be compatible with others that might be merged soon :)

> From my understanding env searches PATH, so using it breaks calling functions with args first, right?

Ah - yes, you're right! I'll think about this a bit :)
```zsh
$ function f() { printenv X }
$ X=100 f
100
$ eval_args X=100 f
env: ‘f’: No such file or directory
```

> Any clear benefit to using regex instead of a glob? I thought it was best to keep it simple, where possible.

I went with the regex to avoid false positives for other args containing an `=`, not that there really _should_ be any (for example, args like `--long-name=value` should always come _after_ the command)

> I'm thinking I'll fixup + force push, since I don't see a reason to keep the original, broken, commit, but I'm still learning the ropes when it comes to open source collaboration. If you have any suggestion, I'll take it.

Yeah - I like to squash commits in PRs to reduce noise in the logs, also :) I can help if you have any issues!

Reply:
this may or may not be a hack, but it appears to work with functions now:

```zsh
function eval_env() {
  local ARG METHOD=
  for ARG in $@; do
    if [[ ${ARG} =~ "^\w+=" ]]; then
      METHOD=env
    else
      if typeset -f ${ARG} > /dev/null; then
        METHOD=eval
      fi
      break
    fi
  done
  if [[ ${METHOD} == "eval" ]]; then
    ${METHOD} "${(q)@}"
  else
    ${METHOD} $@
  fi
}
```

By the way, if you want to avoid a regex, you can use extended globbing/replacement like this for something that should fairly robustly match a variable argument:
```
    if [[ ${ARG#[A-Za-z_]*=} != ${ARG} ]]; then
```
but I think the regex probably clearer :)

Reply:
actually, I believe all you need to do is replace `eval "$*"` in your original code with
```zsh
  eval ${(q)@}
}
```
and it'll work with functions, spaces, environment variables or some combination thereof
(the `${(q)@}` expansion ensures the args are quoted, BTW)

Reply:
there's a _bewildering_ number of variable expansions in Zsh - enjoy ;)
  https://zsh.sourceforge.io/Doc/Release/Expansion.html